packages/grid/x-data-grid-premium/src/hooks/features/cellSelection/useGridCellSelection.ts

```diff
@@ -0,0 +1,363 @@
+    | 'paginationMode'
+  >,
+) => {
+  const visibleRows = useGridVisibleRows(apiRef, props);
+  const lastClickedCell = React.useRef<GridCellCoordinates | null>();
+  const lastMouseDownCell = React.useRef<GridCellCoordinates | null>();
+  const focusedCellWhenShiftWasPressed = React.useRef<GridCellCoordinates | null>(null);
+
+  apiRef.current.registerControlState({
+    stateId: 'cellSelection',
+    propModel: props.cellSelectionModel,
+    propOnChange: props.onCellSelectionModelChange,
+    stateSelector: gridCellSelectionStateSelector,
+    changeEvent: 'cellSelectionChange',
+  });
+
+  const runIfCellSelectionIsEnabled =
+    <Args extends any[]>(callback: (...args: Args) => void) =>
+    (...args: Args) => {
+      if (props.cellSelection) {
+        callback(...args);
+      }
+    };
+
+  const isCellSelected = React.useCallback<GridCellSelectionApi['isCellSelected']>(
+    (id, field) => {
+      if (!props.cellSelection) {
+        return false;
+      }
+      const cellSelectionModel = gridCellSelectionStateSelector(apiRef.current.state);
+      return cellSelectionModel[id] ? !!cellSelectionModel[id][field] : false;
+    },
+    [apiRef, props.cellSelection],
+  );
+
+  const getCellSelectionModel = React.useCallback(() => {
+    return gridCellSelectionStateSelector(apiRef.current.state);
+  }, [apiRef]);
+
+  const setCellSelectionModel = React.useCallback<GridCellSelectionApi['setCellSelectionModel']>(
+    (newModel) => {
+      apiRef.current.setState((prevState) => ({ ...prevState, cellSelection: newModel }));
+      apiRef.current.forceUpdate();
+    },
+    [apiRef],
+  );
+
+  const selectCellRange = React.useCallback<GridCellSelectionApi['selectCellRange']>(
+    (start, end) => {
+      const startRowIndex = apiRef.current.getRowIndexRelativeToVisibleRows(start.id);
+      const startColumnIndex = apiRef.current.getColumnIndex(start.field);
+      const endRowIndex = apiRef.current.getRowIndexRelativeToVisibleRows(end.id);
+      const endColumnIndex = apiRef.current.getColumnIndex(end.field);
+
+      let finalStartRowIndex = startRowIndex;
+      let finalStartColumnIndex = startColumnIndex;
+      let finalEndRowIndex = endRowIndex;
+      let finalEndColumnIndex = endColumnIndex;
+
+      if (finalStartRowIndex > finalEndRowIndex) {
+        finalStartRowIndex = endRowIndex;
+        finalEndRowIndex = startRowIndex;
+      }
+
+      if (finalStartColumnIndex > finalEndColumnIndex) {
+        finalStartColumnIndex = endColumnIndex;
+        finalEndColumnIndex = startColumnIndex;
+      }
+
+      const visibleColumns = apiRef.current.getVisibleColumns();
+      const rowsInRange = visibleRows.rows.slice(finalStartRowIndex, finalEndRowIndex + 1);
+      const columnsInRange = visibleColumns.slice(finalStartColumnIndex, finalEndColumnIndex + 1);
+
+      const newModel = rowsInRange.reduce<GridCellSelectionModel>((acc, row) => {
+        acc[row.id] = columnsInRange.reduce<Record<GridColDef['field'], boolean>>(
```

Comment:
If I understand correctly, the cell selection model keeps track of every selected cell by id + field.
I've noticed a few things related to this:
- you can create a range selection with gaps
- you end up with gaps in range selection after sorting

Maybe we should consider a selection model that has only start and end cell coordinates, where coordinates are row and column indices rather than rowId and column field?

Reply:
Since the majority of use cases for cell range selection don't involve sorting and/or filtering, we decided to keep the existing behavior and reconsider it in the future if needed.